prow/cmd/tackle/ingress.go

```diff
@@ -0,0 +1,81 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"github.com/sirupsen/logrus"
+	extensions "k8s.io/api/extensions/v1beta1"
+	networking "k8s.io/api/networking/v1beta1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/version"
+	"k8s.io/client-go/kubernetes"
+)
+
+// https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/20190125-ingress-api-group.md#116
+// NOTE: GCE ingress controller does not target "networking.k8s.io/v1beta1": https://github.com/kubernetes/ingress-gce/issues/770
+const networkingIngressKubeVersion = "v1.16.0"
+
+var runtimeScheme = runtime.NewScheme()
+
+func init() {
+	extensions.AddToScheme(runtimeScheme)
+	networking.AddToScheme(runtimeScheme)
+}
+
+// networkingIngressAvailable checks if the "networking.k8s.io/v1beta1" resource is available based on Kubernetes version.
+func networkingIngressAvailable(kc kubernetes.Interface) bool {
+	serverVersion, err := kc.Discovery().ServerVersion()
+	if err != nil {
+		logrus.WithError(err).Fatal("unexpected error parsing Kubernetes version")
+		return false
+	}
+
+	return version.CompareKubeAwareVersionStrings(serverVersion.String(), networkingIngressKubeVersion) <= 0
+}
+
+// fromExtensions converts a legacy "extensions/v1beta1" IngressList to the newer "networking.k8s.io/v1beta1" IngressList.
+func fromExtensions(old *extensions.IngressList) (*networking.IngressList, error) {
+	networkingIngress := &networking.IngressList{}
+
+	err := runtimeScheme.Convert(old, networkingIngress, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return networkingIngress, nil
+}
+
+// toIngress normalizes a given IngressList to the newer "networking.k8s.io/v1beta1" type.
+func toIngress(obj interface{}) (*networking.IngressList, bool) {
```

Comment:
The reflection here is unnecessary. As a general rule, stop yourself whenever you find yourself using it and figure out how to avoid it (note that the rest of the repo basically never depends on this)

You know in `ingress()` whether you have a `networking.IngressList` -- in which case call is unnecessary, or else you have an `extensions.IngressList` and definitely need to make this call, so this should be an `extensions.IngressList`, not `interface{}`

Reply:
Those are my JavaScript and Python roots showing through 😉. I will clean this up to be more strict with the typing!